beluga-aws-sdk/src/lib.rs

```diff
@@ -17,45 +20,75 @@ impl TunnelManager {
     ///
     /// This function creates a new [`TunnelManager`] instance by subscribing to
     /// the MQTT topic `$aws/things/{thing_name}/tunnels/notify` with a quality
-    /// of service level of at least once. The resulting
-    /// [`MqttSubscriber`](Subscriber) is stored in the [`TunnelManager`]
-    /// instance.
+    /// of service level of at least once.
     ///
     /// # Arguments
     ///
     /// * `mqtt` - The `MqttClient` instance to use for the MQTT subscription.
     /// * `thing_name` - The name of the AWS IoT thing to subscribe to.
     pub async fn new(mqtt: MqttClient, thing_name: &str) -> Result<Self> {
-        let subscriber = mqtt
+        let mut subscriber = mqtt
             .subscribe(
                 format!("$aws/things/{thing_name}/tunnels/notify").as_str(),
                 QoS::AtLeastOnce,
             )
             .await?;
 
-        Ok(Self { subscriber })
+        let cancel = CancellationToken::new();
+        let cancel_child = cancel.child_token();
+
+        let _handler = tokio::spawn(async move {
+            let mut set = JoinSet::new();
+            loop {
+                tokio::select! {
+                    _ = cancel_child.cancelled() => {
+                        debug!("let's shutdown all the services");
+                        set.shutdown().await;
+                        return Ok(());
+                    }
+                    packet = subscriber.recv() => {
+                        debug!("spawn new service");
+                        set.spawn(service(packet?, cancel_child.clone()));
```

Comment:
Instead of doing `child_token()` outside, it should be done right here.
Basically, the user passes is a cancel token to `new` and we forward child tokens to the `service` function.

Reply:
I think it would be better to have a simple `shutdown` or `cancel` method for this.